RFC-0010-tensor-ref.md

```diff
@@ -0,0 +1,119 @@
+  that the compiler is able to eliminate the `base_` destructor.  This
+  makes it possible to take out a `const Tensor&`, which makes it easier
+  to incrementally convert code to use TensorRef (given a TensorRef, you
+  can always convert it back into `const Tensor&` if you need to call a
+  function that takes the legacy type).
+
+  The downside to this implementation strategy is that TensorRef is
+  not a trivial object (even though, after optimizations, the
+  constructor and destructor are in fact trivial), so that
+  per the Itanium C++ ABI, we must still push this object to the stack
+  and pass it by reference.  So we don't actually manage to eliminate
+  a double indirection here; in fact, if you take a TensorRef and pass
+  it on by value, the generated code will be worse (as each copy
+  construction will push another copy of TensorRef onto the stack).
+  This problem can be fixed by reimplementing TensorRef as directly
+  holding onto a `TensorImpl*`, but then the `const Tensor&` conversion
+  is unimplementable.  Note that if TensorRef is trivial, methods
+  must be implemented carefully to inline away all class calls (e.g.,
+  by delegating onto a out-of-line non-method function), otherwise
+  the same Itanium ABI requirement will force us to put the TensorRef
+  onto the stack.
+
+* Implicit conversion to `const Tensor&` on temporaries is unsafe,
+  because const reference lifetime extension doesn't apply.  Thus, the
+  ref qualifier `&` ensures that this conversion is not permitted on
+  temporaries.  In standard usage, `TensorRef` will be an `lvalue`
+  (e.g., passed in as an argument) and the implicit conversion should
+  always be valid.
+
+* Due to the above, if IValue is adjusted to no longer hold onto a
+  Tensor, then IValue `toTensor` is obligated to return a `TensorRef`
+  by value (because there is no non-temporary `TensorRef` and therefore
+  we cannot manufacture a `const Tensor&`.)  This means that some code
+  that was previously written assuming a `const Tensor&` return may
+  no longer work (e.g., because the ref qualifier rules out the implicit
+  conversion).  However, the author of this RFC does not believe it
+  is possible to introduce unsafety this way.
+
+## Alternatives
+
+* Presently, IValue produces a `const Tensor&` in its interface by
+  directly storing a `Tensor`.  It may be possible to solve include
+  cycles involving IValue in other ways; in particular, by avoiding
+  use of IValue in Tensor APIs, or by avoiding use of IValue in
+  arguments that are defaulted (defaulting can always be reimplemented
+  with N overloads, one per default argument.  This is not a huge burden
+  in the presence of code generation).
+
+* Instead of trying to force the above implementation of `TensorRef` to
+  be used everywhere, it could instead be a utility class used in
+  limited situations to improve interoperability with code that expects
+  a `const Tensor&` when you don't have a `Tensor` available.  The true,
```

Comment:
I don't follow why there are two objects - a trivial and a nontrivial one. Can you give some more details?

Reply:
The nontrivial TensorRef class is given in this proposal. The trivial class is:

```
class TrivialTensorRef {
  TensorImpl* impl_;
};
```

Reply:
oh I see. I like the trivial one much better from a design point of view ;) But I guess you agree on that point and the non-trivial one only exists for the bc reasons discussed above.